File: 76.最小覆盖子串.cpp
/*
 * @lc app=leetcode.cn id=76 lang=cpp
 *
 * [76] 最小覆盖子串
 */
#include<iostream>
#include<vector>
#include<algorithm>
#include<limits.h>
#include<string>
#include<unordered_map>
using namespace std;
// @lc code=start
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> need, window;
        for (char c : t)
            need[c]++;
        int left = 0, right = 0, validNum = 0; // validNum 记录window内包含的t中的字符(种类）数量
        int lenSubstr = INT_MAX, startSubstr = 0;
        while (right < s.length()) {
            char cur = s[right];
            right++;
            if (need.count(cur)) {
                window[cur]++;
                if (window[cur] == need[cur])
                    validNum++;
            }
            while (validNum == need.size()) {
                if (right - left < lenSubstr) {
                    lenSubstr = right - left;
                    startSubstr = left;
                }
                char shrink = s[left];
                left++;
                if (need.count(shrink)) { // 只对t中出现的字符进行判断，否则会得到错误的validNum!
                    window[shrink]--;
                    if (window[shrink] < need[shrink])
                        validNum--;
                }
            }
        }
        return lenSubstr == INT_MAX ? "" : s.substr(startSubstr, lenSubstr);
    }
};
// @lc code=end

File: 76.最小覆盖子串.java

/*
 * @lc app=leetcode.cn id=76 lang=java
 *
 * [76] 最小覆盖子串
 */
import java.util.*;

// @lc code=start
class Solution {
    public String minWindow(String s, String t) {
        HashMap<Character, Integer> need = new HashMap<>(),
                window = new HashMap<>();
        for (int i = 0; i < t.length(); i++) {
            char c = t.charAt(i);
            need.put(c, need.getOrDefault(c, 0) + 1);
        }
        int left = 0, right = 0, validNum = 0;
        int subStrLen = Integer.MAX_VALUE, subStrStart = 0;
        while (right < s.length()) {
            char cur = s.charAt(right);
            right++;
            if (need.containsKey(cur)) {
                window.put(cur, window.getOrDefault(cur, 0) + 1);
                if (window.get(cur).equals(need.get(cur))) {
                    validNum++;
                }
            }
            while (validNum == need.size()) {
                if (right - left < subStrLen) {
                    subStrLen = right - left;
                    subStrStart = left;
                }
                char shrink = s.charAt(left);
                left++;
                if (need.containsKey(shrink)) {
                    if (window.get(shrink).equals(need.get(shrink))) {
                        validNum--;
                    }
                    window.put(shrink, window.get(shrink) - 1);
                }
            }
        }
        return subStrLen == Integer.MAX_VALUE ? "" : s.substring(subStrStart, subStrStart + subStrLen);
    }
}
// @lc code=end
